### ISSUE TITLE
Determine Molecular Equality with wildcard SMARTS

### ISSUE THREAD

**Issue by b-mahjour** [rdkit/rdkit#3117]
How would one go about determining if a molecule matches another molecule with a wildcard:

t1 = Chem.MolFromSmarts("C*")
t2 = Chem.MolFromSmiles("CC")

print(Chem.MolToSmiles(t1)==Chem.MolToSmiles(t2)) (canonical smiles will not work in this instance)
print(t1==t2) (I understand the '==' wouldn't work in this case, but this is what I want)

**Comment by greglandrum** [comment]
Molecular equality isn't a particularly well defined concept.
What are you trying to achieve?

It looks like you want some definition of equality such that Chem.MolFromSmarts('C*') and Chem.MolFromSmiles('CC') are equal, and I'm having a hard time imagining what that means.

**Comment by b-mahjour** [comment]
Thanks for the response!

I was hoping to find a way to determine if I could get all molecules that match that template "C*" or "C*C" and so on. Here are some examples:

"C*": CC, CN, CH
"C*C": COC, CNC, CCC

**Comment by greglandrum** [comment]
That's what substructure matching is for:
http://rdkit.org/docs/GettingStartedInPython.html#substructure-searching

Note that in order to match `CH` you will need to explicitly add Hs to the molecules that you're querying against.

**Comment by b-mahjour** [comment]
Right, but substructure matching would also match CCC (or any molecule containing a carbon) to C*. I only want chemicals matching the direct template. My solution right now is to use substructure matching and check to see if the number of atoms are the same, but I was wondering if there was a specific function that does this exact template matching. 

Thank you!

**Comment by greglandrum** [comment]
I’d do exactly that: check for a substructure match and then make sure the number of atoms and bonds is the same.
You can save time by checking the atom and bond count match first

**Comment by bp-kelley** [comment]
There is another way to do this as well which I'll leave for posterity.

SMARTS supports an explicit degree "[C&D1]" which means a carbon attached to only one heavy atom.  There is an RDKit function adjustQueryParameters that can add these to your queries automatically:

```
params = Chem.AdjustQueryParameters()
params.adjustDegree = True   # match all the bond degress of the query
params.adjustDegreeFlags = 0 # don't ignore any atoms (i.e. dummy)

smiles = ('CC', 'CN', 'CO', 'COC', 'CNC', 'CCC')

queries = "C*", "C*C"

for q in queries:
    print('query', q)
    sma = Chem.MolFromSmarts(q)
    assert sma
    sma2 = Chem.AdjustQueryProperties(sma, params)
    for smi in smiles:
        print(smi, Chem.MolFromSmiles(smi).HasSubstructMatch(sma2))
```

output 

```
query C*
CC True
CN True
CO True
COC False
CNC False
CCC False
query C*C
CC False
CN False
CO False
COC True
CNC True
CCC True
```

This seems also to do what you want.

**Comment by b-mahjour** [comment]
That's fantastic. Thank you so much.
